File: Java/Graphs/dial_algorithm.java
import java.util.*;

public class dial_algorithm {
    static final int INF = Integer.MAX_VALUE;
    private int V;
    private ArrayList<ArrayList<Tuple>> adj;

    public dial_algorithm(int v) {
        this.V = v;
        this.adj = new ArrayList<ArrayList<Tuple>>();
        for (int i = 0; i < v; i++)
            this.adj.add(new ArrayList<Tuple>());
    }

    public void AddEdge(int u, int v, int w) {
        adj.get(u).add(new Tuple(v, w));
        adj.get(v).add(new Tuple(u, w));
    }

    public void shortestPath(int src, int W) {
        int[] dist = new int[V];
        Arrays.fill(dist, INF);
        ArrayList<Integer>[] B = new ArrayList[W * V + 1];
        for (int i = 0; i < W * V + 1; i++)
            B[i] = new ArrayList<Integer>();
        B[0].add(src);
        dist[src] = 0;
        int idx = 0;
        while (true) {
            while (B[idx].size() == 0 && idx < W * V)
                idx++;
            if (idx == W * V)
                break;
            int u = B[idx].get(0);
            B[idx].remove(0);
            for (Tuple i : adj.get(u)) {
                int v = i.v;
                int weight = i.w;
                int du = dist[u];
                int dv = dist[v];
                if (dv > du + weight) {
                    dist[v] = du + weight;
                    dv = dist[v];
                    B[dv].add(0, v);
                }
            }
        }
        System.out.println("Distance from Source");
        for (int i = 0; i < V; ++i)
            System.out.println(i + "\t\t" + dist[i]);
    }

    static class Tuple {
        int v, w;

        Tuple(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    public static void main(String[] args) {
        int V = 9;
        dial_algorithm g = new dial_algorithm(V);
        g.AddEdge(0, 1, 4);
        g.AddEdge(0, 7, 8);
        g.AddEdge(1, 2, 8);
        g.AddEdge(1, 7, 11);
        g.AddEdge(2, 3, 7);
        g.AddEdge(2, 8, 2);
        g.AddEdge(2, 5, 4);
        g.AddEdge(3, 4, 9);
        g.AddEdge(3, 5, 14);
        g.AddEdge(4, 5, 10);
        g.AddEdge(5, 6, 2);
        g.AddEdge(6, 7, 1);
        g.AddEdge(6, 8, 6);
        g.AddEdge(7, 8, 7);
        g.shortestPath(0, 14);
    }
}

File: Python/Graphs/dial_algorithm.py
from typing import List, Tuple
INF = 0x3f3f3f3f
class Graph:
	def __init__(self, V: int):
		self.V = V
		self.adj = [[] for _ in range(V)]
	def addEdge(self, u: int, v: int, w: int):
		self.adj[u].append((v, w))
		self.adj[v].append((u, w))
	def shortestPath(self, src: int, W: int):
		dist = [[INF, None] for _ in range(self.V)]
		dist[src][0] = 0
		B = [[] for _ in range(W * self.V + 1)]
		B[0].append(src)
		idx = 0
		while True:
			while len(B[idx]) == 0 and idx < W * self.V:
				idx += 1
			if idx == W * self.V:
				break
			u = B[idx][0]
			B[idx].pop(0)
			for v, weight in self.adj[u]:
				du = dist[u][0]
				dv = dist[v][0]
				if dv > du + weight:
					if dv != INF:
						B[dv].remove(v)
					dist[v][0] = du + weight
					dv = dist[v][0]
					B[dv].append(v)
					dist[v][1] = len(B[dv]) - 1
		print("Distance from Source")
		for i in range(self.V):
			print(f"{i}	 {dist[i][0]}")
def main():
	V = 9
	W = 14
	g = Graph(V)
	g.addEdge(0, 1, 4)
	g.addEdge(0, 7, 8)
	g.addEdge(1, 2, 8)
	g.addEdge(1, 7, 11)
	g.addEdge(2, 3, 7)
	g.addEdge(2, 8, 2)
	g.addEdge(2, 5, 4)
	g.addEdge(3, 4, 9)
	g.addEdge(3, 5, 14)
	g.addEdge(4, 5, 10)
	g.addEdge(5, 6, 2)
	g.addEdge(6, 7, 1)
	g.addEdge(6, 8, 6)
	g.addEdge(7, 8, 7)
	g.shortestPath(0, W)
if __name__ == "__main__":
	main()
